run/tesla_api.py

```diff
@@ -142,16 +142,43 @@ def _get_api_token():
         # case, we can't accurately determine the age of the token, so we just
         # use it. Later executions of the script should run after the date has
         # updated correctly, at which point we can properly compare the dates.
+        max_retries = 24
+        retry_count = 0
+        retry_interval_s = 5
+        cutoff_year = 2019
+        # Wait for system time to be synchronized, up to max_retries times
+        while retry_count < max_retries:
+            now = datetime.now()
+            if now.year >= cutoff_year:  # Time is correct, break the loop
+                break
+            _log(
+                f"System time is out of sync: {now}. Retrying in {retry_interval_s} seconds... (Attempt {retry_count + 1}/{max_retries})")
+            time.sleep(retry_interval_s)  # Sleep for 5 seconds before checking again
+            retry_count += 1
+
         now = datetime.now()
-        if now.year < 2019: # This script was written in 2019.
+        if now.year < cutoff_year:  # This script was written in 2019.
+            _log("Time is still wrong after waiting. Using existing access token.")
+            return tesla_api_json['access_token']
+
+        if not SETTINGS.get('refresh_token'):
+            _log('No refreshing token available. Using existing access token.')
             return tesla_api_json['access_token']
 
         tesla = teslapy.Tesla(SETTINGS['tesla_email'], None)
-        if SETTINGS['REFRESH_TOKEN'] or 0 < tesla.expires_at < time.time():
-            _log('Refreshing api token')
+        # For some reason, the `expires_at` timestamp doesn't exactly match the `exp` field in the JWT body.
+        # The `expires_at` is typically about 1 minute ahead of the `exp` value in the JWT payload.
+        # Access token usually expires after a few hours, so refresh the token if it has less than `expiration_buffer_s` seconds remaining.
+        expiration_buffer_s = 60 * 30
+        if tesla.expires_at <= time.time() - expiration_buffer_s:
+            _log("Refreshing expired access token...")
+            tesla.token['refresh_token'] = SETTINGS.get('refresh_token')
             tesla.refresh_token()
-            tesla_api_json['access_token'] = tesla.token.get('access_token')
```

Comment:
The `access_token` value from teslapy was never synced back to the tesla_api.json file. The `teslapy` uses `cache.json` file to save its own tokens. 

The end result is the token in `tesla_api_json['access_token']` already expired, but the token in `tesla.token.get('access_token')` is ok after refresh. It causes the 401 expired token error.